scripts/start.js

```diff
@@ -119,35 +151,63 @@ compiler.plugin('done', function (stats) {
     console.log('Use ' + chalk.yellow('// eslint-disable-next-line') + ' to ignore the next line.');
     console.log('Use ' + chalk.yellow('/* eslint-disable */') + ' to ignore all warnings in a file.');
   }
-});
+}
 
-function openBrowser() {
-  if (process.platform === 'darwin') {
-    try {
-      // Try our best to reuse existing tab
-      // on OS X Google Chrome with AppleScript
-      execSync('ps cax | grep "Google Chrome"');
-      execSync(
-        'osascript ' +
-        path.resolve(__dirname, './openChrome.applescript') +
-        ' http://localhost:3000/'
-      );
-      return;
-    } catch (err) {
-      // Ignore errors.
-    }
-  }
-  // Fallback to opn
-  // (It will always open new tab)
-  opn('http://localhost:3000/');
+function exponentialBackoff(step) {
+  return Math.pow(2, step);
 }
 
-new WebpackDevServer(compiler, {
+const clientCompiler = webpack(configClient, handleCompileClient);
+clientCompiler.plugin('invalid', webpackOnInvalid);
+clientCompiler.plugin('done', webpackOnDone);
+
+// Here's the server compiler
+webpack(configServer, function(error, stats) {
+  webpackOnDone(stats);
+
+  if (!server) {
+    server = new SingleChild('node', ['build/server/server-dev.js'], {
+      stdio: [0, 1, 2]
+    });
+    server.start();
+  } else {
+    server.restart();
+  }
+});
+
+const devServer = new WebpackDevServer(clientCompiler, {
   historyApiFallback: true,
   hot: true, // Note: only CSS is currently hot reloaded
-  publicPath: config.output.publicPath,
+  publicPath: configClient.output.publicPath,
   quiet: true
-}).listen(3000, function (err, result) {
+});
+
+devServer.use('/', (req, res) => {
```

Comment:
This is the most important part.

We've always had many issues with `webpack-dev-middleware` and from our production experience this simple proxy has always worked pretty well.

Webpack dev server serves just for client assets and HMR provider and everything else (initial renders / custom API handlers etc.) is served through plain old express server.

This proxy automatically exposes all the requests on single port so that it's totally transparent for the user.

Retry back-off algorithm is mandatory since it may happen that HMR request can be triggered before the server is actually compiled.